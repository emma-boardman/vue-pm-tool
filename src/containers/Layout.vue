<template>
  <div>
    <layout-modal
      :if="isAddTaskFormShowing"
      :showModal="isAddTaskFormShowing"
      @closeModal="handleTaskAddModalClose"
    >
      <layout-task-add
        class="modal-task-add"
        :projects="this.projectsList"
        :resources="this.resourceList"
        @submitNewTask="handleFormSubmission"
        @handleEstimate="fetchResourceSchedule"
        @closeModal="handleTaskAddModalClose"
        :availableTimes="{startTime: this.firstAvailableStartTime, endTime: this.firstAvailableEndTime}"
      ></layout-task-add>
    </layout-modal>
    <layout-modal
      :if="isTaskDetailShowing"
      @closeModal="handleTaskDetailModalClose"
      :showModal="isTaskDetailShowing"
      class="modal-task-detail"
    >
    <layout-task-detail :task="taskDetail" @closeModal="handleTaskDetailModalClose"></layout-task-detail>
    </layout-modal>
    <layout-header></layout-header>
    <layout-calendar></layout-calendar>
    <layout-footer></layout-footer>
  </div>
</template>

<script>
import Header from "../components/Header/Header.vue";
import Calendar from "../components/Calendar/Calendar.vue";
import Footer from "../components/Footer/Footer.vue";
import Modal from "../components/UI/Modal/Modal.vue";
import TaskAdd from "../components/Task/TaskAdd/TaskAdd.vue";
import TaskDetail from "../components/Task/TaskDetail/TaskDetail.vue";
import { EventBus } from "../event-bus.js";
import { store } from "../utils/store.js";
import axios from "axios";
import { RepositoryFactory } from "../utils/RepositoryFactory";
const ResourceRepository = RepositoryFactory.get("resources");
const TaskRepository = RepositoryFactory.get("tasks");

export default {
  data: function() {
    return {
      store,
      isResourceTasksLoading: store.state.isResourceTasksLoading,
      isTaskFormPresetsLoading: store.state.isTaskFormPresetsLoading,
      projectsList: store.state.projectsList,
      resourceList: store.state.resourceList,
      taskDetail: {},
      resourceSchedule: {},
      firstAvailableStartTime: "",
      firstAvailableEndTime: ""
    };
  },
  computed: {
    isAddTaskFormShowing: function() {
      return store.state.modalControls.isAddTaskFormShowing;
     },
     isTaskDetailShowing: function() {
       return store.state.modalControls.isTaskDetailShowing;
     },
  },
  created() {
    EventBus.$on("showTaskDetails", this.handleTaskDetailsModalOpen);
    store.fetchResourceTasks();
    this.fetchTaskFormPresets();
  },
  components: {
    layoutHeader: Header,
    layoutCalendar: Calendar,
    layoutFooter: Footer,
    layoutModal: Modal,
    layoutTaskAdd: TaskAdd,
    layoutTaskDetail: TaskDetail
  },
  methods: {
    async fetchTaskFormPresets() {
      this.isTaskFormPresetsLoading = true;
      const { data } = await TaskRepository.getTaskFormPresets();
      console.log(data);
      this.isTaskFormPresetsLoading = false;
      this.projectsList = data["clientProjects"];
      this.resourceList = data["resources"];
      this.handleTaskPresets();
    },
    async fetchResourceSchedule(resourceAndEstimate) {
      this.isResourceScheduleLoading = true;
      const { data } = await ResourceRepository.getResourceSchedule(
        resourceAndEstimate.resourceId
      );
      this.isResourceScheduleLoading = false;
      this.resourceSchedule = data;
      this.handleResourceAvailability(resourceAndEstimate);
    },
    handleTaskPresets: function() {
      let projectsArray = this.projectsList.reduce(function(
        formPresetsArray,
        clientObj
      ) {
        for (let j = 0; j < clientObj.projects.length; j++) {
          let clientObject = {
            clientName: clientObj.clientName,
            projectName: clientObj.projects[j].projectTitle,
            projectId: clientObj.projects[j].projectId
          };
          formPresetsArray.push(clientObject);
        }
        return formPresetsArray;
      },
      []);
      this.projectsList = projectsArray;
    },
    handleTaskAddModalClose: function() {
      store.hideAddTaskForm();
    },
    handleTaskAddModalOpen: function() {
      this.showTaskAdd = true;
    },
    handleTaskDetailsModalOpen: function(task) {
      store.state.modalControls.isTaskDetailShowing = true;
      this.taskDetail = task;
    },
    handleTaskDetailModalClose: function() {
      store.hideTaskDetail();
    },
    handleFormSubmission: function(task) {
      console.log(task);
      this.handleTaskAddModalClose();
      this.postNewTask(task);
      // this.firstAvailableStartTime ='';
      // this.firstAvailableEndTime ='';
    },
    handleResourceAvailability: function(resourceAndEstimate) {
      const weeklyAvailability = Array(45).fill(true);
      const weeklyTimeSlots = [
        "Mon0900",
        "Mon1000",
        "Mon1100",
        "Mon1200",
        "Mon1300",
        "Mon1400",
        "Mon1500",
        "Mon1600",
        "Mon1700",
        "Tues0900",
        "Tues1000",
        "Tues1100",
        "Tues1200",
        "Tues1300",
        "Tues1400",
        "Tues1500",
        "Tues1600",
        "Tues1700",
        "Wed0900",
        "Wed1000",
        "Wed1100",
        "Wed1200",
        "Wed1300",
        "Wed1400",
        "Wed1500",
        "Wed1600",
        "Wed1700",
        "Thurs0900",
        "Thurs1000",
        "Thurs1100",
        "Thurs1200",
        "Thurs1300",
        "Thurs1400",
        "Thurs1500",
        "Thurs1600",
        "Thurs1700",
        "Fri0900",
        "Fri1000",
        "Fri1100",
        "Fri1200",
        "Fri1300",
        "Fri1400",
        "Fri1500",
        "Fri1600",
        "Fri1700"
      ];
      if (this.resourceSchedule.length > 0) {
        for (let n = 0; n < this.resourceSchedule.length; n++) {
          const startTime = this.resourceSchedule[n].taskStartTime;
          const estimatedTime = this.resourceSchedule[n].taskEstimate;
          for (let i = 0; i < weeklyTimeSlots.length; i++) {
            if (weeklyTimeSlots[i] == startTime) {
              for (let j = 0; j < estimatedTime; j++) {
                weeklyAvailability[i + j] = false;
              }
            }
          }
        }
      }
      console.log(weeklyAvailability);
      const estimateArray = Array(resourceAndEstimate.taskEstimate).fill(true);
      let indexOfFirstAvailability = this.findAvailability(
        weeklyAvailability,
        estimateArray
      );
      this.firstAvailableStartTime = weeklyTimeSlots[indexOfFirstAvailability];
      const endTime =
        parseInt(indexOfFirstAvailability) +
        parseInt(resourceAndEstimate.taskEstimate);
      this.firstAvailableEndTime = weeklyTimeSlots[endTime];
    },
    findAvailability(arr, subarr) {
      for (var i = 0; i < 1 + (arr.length - subarr.length); i++) {
        var j = 0;
        for (; j < subarr.length; j++) if (arr[i + j] !== subarr[j]) break;
        if (j == subarr.length) return i;
      }
      return -1;
    }
  }
};
</script>

<style scoped>
.modal-task-detail >>> .modal {
  padding: 0;
  top: 10%;
  border: var(--dark-grey) solid 1px;
}
</style>
