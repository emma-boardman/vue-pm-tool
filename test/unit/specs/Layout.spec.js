import { mount } from '@vue/test-utils'
import Layout from '../../../src/containers/Layout'


describe('Layout', () => {
    let wrapper 

    beforeEach(() => {
        wrapper = mount(Layout)
    })

    test("component is rendered", () => {
        expect(wrapper.element).toMatchSnapshot();
    })

    test("handleResourceAvailability returns first available timeslot", () => {
        wrapper.vm.handleResourceAvailability({
            resourceId: '2',
            taskEstimate: '4'
        })
        expect(wrapper.vm.firstAvailableStartTime).toEqual('Tues1300');
        expect(wrapper.vm.firstAvailableEndTime).toEqual('Tues1700');
    })
})