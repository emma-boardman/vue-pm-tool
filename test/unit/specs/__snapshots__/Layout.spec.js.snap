// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component is rendered 1`] = `
<div>
  <div>
    <div
      class="backdrop"
      style="display: none;"
    />
     
    <div
      class="modal modal--hide"
    >
      <div
        class="add-task modal-task-add"
      >
        <h2>
          Add New Task
        </h2>
         
        <form>
          <div
            class="top-on-desktop"
          >
            <div
              class="left-on-desktop"
            >
              <ul
                class="form-wrapper"
              >
                <li
                  class="form-row"
                >
                  <label
                    for="projectId"
                  >
                    Project
                  </label>
                   
                  <select
                    id="taskProject"
                    required="required"
                  />
                </li>
                 
                <li
                  class="form-row"
                >
                  <label
                    for="taskTitle"
                  >
                    Task Title
                  </label>
                   
                  <input
                    id="taskTitle"
                    required="required"
                    type="text"
                  />
                </li>
                 
                <li
                  class="form-row"
                >
                  <label
                    for="taskArea"
                  >
                    Affected area
                  </label>
                   
                  <input
                    id="taskArea"
                    required="required"
                    type="text"
                  />
                </li>
                 
                <li
                  class="form-row"
                >
                  <label
                    for="taskErroneousBehaviour"
                  >
                    Erroneous behaviour
                  </label>
                   
                  <textarea
                    cols="20"
                    id="taskErroneousBehaviour"
                    required="required"
                    rows="2"
                  />
                </li>
                 
                <li
                  class="form-row"
                >
                  <label
                    for="taskExpectedBehaviour"
                  >
                    Expected behaviour
                  </label>
                   
                  <textarea
                    cols="20"
                    id="taskExpectedBehaviour"
                    required="required"
                    rows="2"
                  />
                </li>
              </ul>
            </div>
             
            <div
              class="right-on-desktop"
            >
              <ul
                class="form-wrapper"
              >
                <li
                  class="form-row"
                >
                  <label
                    for="taskImpact"
                  >
                    Impact on business
                  </label>
                   
                  <select
                    id="taskImpact"
                    required="required"
                  >
                    <option>
                      Low
                    </option>
                    <option>
                      Medium
                    </option>
                    <option>
                      High
                    </option>
                  </select>
                </li>
                 
                <li
                  class="form-row"
                >
                  <label
                    for="taskTimeNoticed"
                  >
                    Time noticed
                  </label>
                   
                  <input
                    id="taskTimeNoticed"
                    required="required"
                    type="text"
                  />
                </li>
                 
                <li
                  class="form-row"
                >
                  <label
                    for="taskRecentChanges"
                  >
                    Recent Changes
                  </label>
                   
                  <input
                    id="taskRecentChanges"
                    required="required"
                    type="text"
                  />
                </li>
                 
                <li
                  class="form-row"
                >
                  <label
                    for="resourceId"
                  >
                    Resource
                  </label>
                   
                  <select
                    id="resourceId"
                    required="required"
                  />
                </li>
                 
                <li
                  class="form-row"
                >
                  <label
                    for="taskEstimate"
                  >
                    Estimate (hours)
                  </label>
                   
                  <input
                    id="taskEstimate"
                    required="required"
                    type="number"
                  />
                </li>
                 
                <li
                  class="form-row"
                >
                  <label
                    for="taskSchedule"
                  >
                    First Available Time Slot
                  </label>
                   
                  <div>
                    <div>
                      <p>
                        Requires resource and estimate input
                      </p>
                    </div>
                  </div>
                </li>
              </ul>
            </div>
          </div>
           
          <div
            class="bottom-on-desktop"
          >
            <button
              class="add-task-btn"
              type="submit"
            >
              Add Task
            </button>
             
            <p
              class="cancel-btn"
            >
              Cancel
            </p>
          </div>
        </form>
      </div>
    </div>
  </div>
   
  <div
    class="modal-task-detail"
  >
    <div
      class="backdrop"
      style="display: none;"
    />
     
    <div
      class="modal modal--hide"
    >
      <div
        class="task-detail"
      >
        <div
          class="task-detail__top-bar"
          style="background-color: rgb(29, 164, 193);"
        >
          <div>
            
          </div>
           
          <div
            class="task-detail__top-bar__close-btn"
          >
            x
          </div>
        </div>
         
        <div
          class="task-detail__task-content"
        >
          <p
            class="task-detail__task-content__task-title"
          >
            
      
      
            <span
              class="task-detail__task-content--impact"
            >
              
            </span>
          </p>
           
          <dl>
            <dt>
              Project
            </dt>
             
            <dd>
              
            </dd>
             
            <dt>
              Error
            </dt>
             
            <dd>
              
            </dd>
             
            <dt>
              Scheduled for
            </dt>
             
            <dd>
               - 
            </dd>
          </dl>
        </div>
      </div>
    </div>
  </div>
   
  <header>
    <h1>
      Company X
    </h1>
  </header>
   
  <div
    class="scrolling-wrapper"
  >
    <div
      class="wrapper"
    >
      <div
        class="row-headers"
      >
        <div
          class="empty"
        />
         
        <div
          class="weekly-grid"
        >
          <div
            class="day-date weekday monday"
          >
            <div
              class="day"
            >
              <p>
                Monday 8th July
              </p>
            </div>
             
            <div
              class="hours"
            >
              <div
                class="hour"
              >
                9
              </div>
               
              <div
                class="hour"
              >
                10
              </div>
               
              <div
                class="hour"
              >
                11
              </div>
               
              <div
                class="hour"
              >
                12
              </div>
               
              <div
                class="hour"
              >
                13
              </div>
               
              <div
                class="hour"
              >
                14
              </div>
               
              <div
                class="hour"
              >
                15
              </div>
               
              <div
                class="hour"
              >
                16
              </div>
               
              <div
                class="hour"
              >
                17
              </div>
            </div>
          </div>
          <div
            class="day-date weekday tuesday"
          >
            <div
              class="day"
            >
              <p>
                Tuesday 9th July
              </p>
            </div>
             
            <div
              class="hours"
            >
              <div
                class="hour"
              >
                9
              </div>
               
              <div
                class="hour"
              >
                10
              </div>
               
              <div
                class="hour"
              >
                11
              </div>
               
              <div
                class="hour"
              >
                12
              </div>
               
              <div
                class="hour"
              >
                13
              </div>
               
              <div
                class="hour"
              >
                14
              </div>
               
              <div
                class="hour"
              >
                15
              </div>
               
              <div
                class="hour"
              >
                16
              </div>
               
              <div
                class="hour"
              >
                17
              </div>
            </div>
          </div>
          <div
            class="day-date weekday wednesday"
          >
            <div
              class="day"
            >
              <p>
                Wednesday 10th July
              </p>
            </div>
             
            <div
              class="hours"
            >
              <div
                class="hour"
              >
                9
              </div>
               
              <div
                class="hour"
              >
                10
              </div>
               
              <div
                class="hour"
              >
                11
              </div>
               
              <div
                class="hour"
              >
                12
              </div>
               
              <div
                class="hour"
              >
                13
              </div>
               
              <div
                class="hour"
              >
                14
              </div>
               
              <div
                class="hour"
              >
                15
              </div>
               
              <div
                class="hour"
              >
                16
              </div>
               
              <div
                class="hour"
              >
                17
              </div>
            </div>
          </div>
          <div
            class="day-date weekday thursday"
          >
            <div
              class="day"
            >
              <p>
                Thursday 11th July
              </p>
            </div>
             
            <div
              class="hours"
            >
              <div
                class="hour"
              >
                9
              </div>
               
              <div
                class="hour"
              >
                10
              </div>
               
              <div
                class="hour"
              >
                11
              </div>
               
              <div
                class="hour"
              >
                12
              </div>
               
              <div
                class="hour"
              >
                13
              </div>
               
              <div
                class="hour"
              >
                14
              </div>
               
              <div
                class="hour"
              >
                15
              </div>
               
              <div
                class="hour"
              >
                16
              </div>
               
              <div
                class="hour"
              >
                17
              </div>
            </div>
          </div>
          <div
            class="day-date weekday friday"
          >
            <div
              class="day"
            >
              <p>
                Friday 12th July
              </p>
            </div>
             
            <div
              class="hours"
            >
              <div
                class="hour"
              >
                9
              </div>
               
              <div
                class="hour"
              >
                10
              </div>
               
              <div
                class="hour"
              >
                11
              </div>
               
              <div
                class="hour"
              >
                12
              </div>
               
              <div
                class="hour"
              >
                13
              </div>
               
              <div
                class="hour"
              >
                14
              </div>
               
              <div
                class="hour"
              >
                15
              </div>
               
              <div
                class="hour"
              >
                16
              </div>
               
              <div
                class="hour"
              >
                17
              </div>
            </div>
          </div>
           
        </div>
      </div>
       
      <div
        class="resources"
      />
    </div>
  </div>
   
  <footer>
    <div
      class="add-button-wrapper"
    >
      <img
        src="../../../assets/addButton.svg"
      />
    </div>
  </footer>
</div>
`;
